test/move_effect_fling.c

```diff
@@ -0,0 +1,336 @@
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_RECYCLE].effect == EFFECT_RECYCLE);
+        PLAYER(SPECIES_WOBBUFFET) {Item(ITEM_RAZOR_CLAW); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(player, MOVE_FLING);}
+        TURN { MOVE(player, MOVE_RECYCLE);}
+        TURN { MOVE(player, MOVE_FLING);}
+    } SCENE {
+        MESSAGE("Wobbuffet used Fling!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_FLING, player);
+        HP_BAR(opponent);
+        MESSAGE("Wobbuffet used Recycle!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_RECYCLE, player);
+        MESSAGE("Wobbuffet found one Razor Claw!");
+        MESSAGE("Wobbuffet used Fling!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_FLING, player);
+        HP_BAR(opponent);
+    }
+}
+
+SINGLE_BATTLE_TEST("Fling - Item is lost even when there is no target")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_SELF_DESTRUCT].effect == EFFECT_EXPLOSION);
+        PLAYER(SPECIES_WOBBUFFET) {Item(ITEM_RAZOR_CLAW); Speed(2); }
+        OPPONENT(SPECIES_WOBBUFFET) {Speed(5); }
+        OPPONENT(SPECIES_WOBBUFFET) {Speed(5); }
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_SELF_DESTRUCT); MOVE(player, MOVE_FLING); SEND_OUT(opponent, 1); }
+        TURN { MOVE(player, MOVE_FLING); }
+    } SCENE {
+        MESSAGE("Foe Wobbuffet used SelfDestruct!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_SELF_DESTRUCT, opponent);
+        HP_BAR(player);
+        MESSAGE("Foe Wobbuffet fainted!");
+        MESSAGE("Wobbuffet used Fling!");
+        MESSAGE("But it failed!");
+
+        MESSAGE("Wobbuffet used Fling!");
+        MESSAGE("But it failed!");
+    } THEN {
+        EXPECT_EQ(player->item, ITEM_NONE);
+    }
+}
+
+SINGLE_BATTLE_TEST("Fling - Item is lost when target protects itself")
+{
+    GIVEN {
+        ASSUME(gBattleMoves[MOVE_PROTECT].effect == EFFECT_PROTECT);
+        PLAYER(SPECIES_WOBBUFFET) {Item(ITEM_RAZOR_CLAW); }
+        OPPONENT(SPECIES_WOBBUFFET);
+    } WHEN {
+        TURN { MOVE(opponent, MOVE_PROTECT); MOVE(player, MOVE_FLING);}
+        TURN { MOVE(player, MOVE_FLING); }
+    } SCENE {
+        MESSAGE("Foe Wobbuffet used Protect!");
+        ANIMATION(ANIM_TYPE_MOVE, MOVE_PROTECT, opponent);
+        MESSAGE("Wobbuffet used Fling!");
+        MESSAGE("Foe Wobbuffet protected itself!");
+
+        MESSAGE("Wobbuffet used Fling!");
+        MESSAGE("But it failed!");
+    } THEN {
+        EXPECT_EQ(player->item, ITEM_NONE);
+    }
+}
+
+SINGLE_BATTLE_TEST("Fling doesn't consume the item if pokemon is asleep/frozen/paralyzed")
+{
+    u32 status;
+    u16 item;
+
+    PARAMETRIZE {status = STATUS1_SLEEP_TURN(2); item = ITEM_RAZOR_CLAW; }
+    PARAMETRIZE {status = STATUS1_PARALYSIS; item = ITEM_RAZOR_CLAW; }
+    PARAMETRIZE {status = STATUS1_FREEZE; item = ITEM_RAZOR_CLAW; }
+    PARAMETRIZE {status = STATUS1_SLEEP_TURN(2); item = ITEM_NONE; }
+    PARAMETRIZE {status = STATUS1_PARALYSIS; item = ITEM_NONE; }
+    PARAMETRIZE {status = STATUS1_FREEZE; item = ITEM_NONE; }
```

Comment:
I think the `item` field here is unneeded, since you already made a test for failing when the item is `ITEM_NONE`
```suggestion
    PARAMETRIZE { status = STATUS1_SLEEP_TURN(2); item = ITEM_RAZOR_CLAW; }
    PARAMETRIZE { status = STATUS1_PARALYSIS; item = ITEM_RAZOR_CLAW; }
    PARAMETRIZE { status = STATUS1_FREEZE; item = ITEM_RAZOR_CLAW; }
    PARAMETRIZE { status = STATUS1_SLEEP_TURN(2); item = ITEM_NONE; }
    PARAMETRIZE { status = STATUS1_PARALYSIS; item = ITEM_NONE; }
    PARAMETRIZE { status = STATUS1_FREEZE; item = ITEM_NONE; }
```